Assets/Shopify/UIToolkit/Editor/ProductPicker.cs

```diff
@@ -0,0 +1,155 @@
+﻿namespace Shopify.UIToolkit.Editor {
+    using System.Collections;
+    using System.Collections.Generic;
+    using UnityEngine;
+    using UnityEditor;
+    using Shopify.Unity;
+    using Shopify.Unity.SDK;
+    using System;
+    using System.Linq;
+
+    public interface IProductPickerView {
+        void DrawProductLoadingState(SerializedObject serializedObject);
+        void DrawProductPicker(SerializedObject serializedObject);
+        void DrawProductLoadingError(SerializedObject serializedObject);
+        void DrawShopHasNoProductsState(SerializedObject serializedObject);
+    }
+
+    public class ProductPicker : IProductPickerView {
+        private class ProductNameToGIDMapEntry {
+            public string ProductName;
+            public string ProductGID;
+        }
+
+        public IProductPickerView View;
+        private static List<ProductNameToGIDMapEntry> _cachedProductNameToGIDMap;
+        private List<ProductNameToGIDMapEntry> _productNameToGIDMap;
+        private ShopifyClient _client;
+        private ShopifyError _error;
+
+        public bool ProductsFinishedLoading() {
+            return _productNameToGIDMap != null || _error != null;
+        }
+
+        public ProductPicker(ShopifyClient client) {
+            _client = client;
+            View = this;
+            LoadProducts();
+        }
+
+        public void DrawInspectorGUI(SerializedObject serializedObject) {
+            if (_error != null) {
+                View.DrawProductLoadingError(serializedObject);
+                return;
+            }
+
+            if (!ProductsFinishedLoading()) {
+                View.DrawProductLoadingState(serializedObject);
+                return;
+            }
+
+            if (_productNameToGIDMap.Count == 0) {
+                View.DrawShopHasNoProductsState(serializedObject);
+                return;
+            }
+
+            View.DrawProductPicker(serializedObject);
+        }
+
+        public void LoadProducts() {
+            if (_cachedProductNameToGIDMap != null) {
+                _productNameToGIDMap = _cachedProductNameToGIDMap;
+            }
+
+            _client.products((products, error) => {
+                if (error != null) {
+                    OnFailedToLoadProducts(error);
+                    return;
+                }
+
+                OnLoadedProducts(products);
+            });
+        }
+
+        public void RefreshProducts() {
+            _error = null;
+            _productNameToGIDMap = null;
+            _cachedProductNameToGIDMap = null;
+            LoadProducts();
+        }
+
+        private void OnFailedToLoadProducts(ShopifyError error) {
+            _error = error;
+            _cachedProductNameToGIDMap = null;
```

Comment:
Just because we failed doesn't mean we should invalidate previously cached products no?